File: sdks/node/src/resources/billing.ts
/**
 * HookSniff SDK — Billing Resource
 */

import { HttpMethod, HookSniffRequest, type HookSniffRequestContext } from "../request";
import type {
  SubscriptionResponse,
  UpgradeRequest,
  UpgradeResponse,
  UsageResponse,
  InvoiceResponse,
  BillingPortalResponse,
} from "../models";

export class Billing {
  constructor(private readonly requestCtx: HookSniffRequestContext) {}

  /** Get current subscription. */
  public subscription(): Promise<SubscriptionResponse> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/billing/subscription");
    return request.send(this.requestCtx, (json) => json as SubscriptionResponse);
  }

  /** Get usage for current billing period. */
  public usage(): Promise<UsageResponse> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/billing/usage");
    return request.send(this.requestCtx, (json) => json as UsageResponse);
  }

  /** Get invoices. */
  public invoices(): Promise<InvoiceResponse[]> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/billing/invoices");
    return request.send(this.requestCtx, (json) => json as InvoiceResponse[]);
  }

  /** Upgrade plan. Returns checkout URL. */
  public upgrade(body: UpgradeRequest): Promise<UpgradeResponse> {
    const request = new HookSniffRequest(HttpMethod.POST, "/v1/billing/upgrade");
    request.setBody(body);
    return request.send(this.requestCtx, (json) => json as UpgradeResponse);
  }

  /** Open billing portal (Polar.sh/Stripe). */
  public portal(): Promise<BillingPortalResponse> {
    const request = new HookSniffRequest(HttpMethod.POST, "/v1/billing/portal");
    return request.send(this.requestCtx, (json) => json as BillingPortalResponse);
  }
}

File: sdks/rust/src/resources/billing.rs
//! HookSniff API Resource: Billing
//!
//! Billing and subscription management.

use crate::request::{HookSniffRequest, HookSniffRequestContext, HttpMethod};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct PlanOutput {
    pub plan: String,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub webhook_limit: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub endpoint_limit: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_usage: Option<u64>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpgradeInput {
    pub plan: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PortalOutput {
    pub url: String,
}

#[derive(Debug)]
pub struct Billing {
    ctx: HookSniffRequestContext,
}

impl Billing {
    pub fn new(ctx: HookSniffRequestContext) -> Self {
        Self { ctx }
    }

    /// Get current plan
    pub fn plan(&self) -> Result<PlanOutput, Box<dyn std::error::Error>> {
        let req = HookSniffRequest::new(HttpMethod::Get, "/v1/billing/plan");
        req.send(&self.ctx)
    }

    /// Upgrade plan
    pub fn upgrade(&self, input: &UpgradeInput) -> Result<PlanOutput, Box<dyn std::error::Error>> {
        let mut req = HookSniffRequest::new(HttpMethod::Post, "/v1/billing/upgrade");
        req.set_body(input);
        req.send(&self.ctx)
    }

    /// Get billing portal URL
    pub fn portal(&self) -> Result<PortalOutput, Box<dyn std::error::Error>> {
        let req = HookSniffRequest::new(HttpMethod::Post, "/v1/billing/portal");
        req.send(&self.ctx)
    }
}
